### PR TITLE
Issue 813 event list

### PR DESCRIPTION
**Pull request by qasema**
- The event list should show upcoming events first then past events come after in sorted order. 
- past events should be grayed out so it shows that they happened in the past 
- only admins and student staff should see the past events 

*Test* 
- got to the event list and view the list of created events 
- you can create new events to test it 

fixes issue  #813

### CODE REVIEW

File: app/templates/events/event_list.html [hunk 1]
```diff
@@ -69,22 +70,62 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
+                {% else %}
+                  <td  class=text-secondary>This event does not belong to any program.</td>
+                {% endif %}
+              {% endif %}
+              {% if user.isCeltsAdmin or user.isProgramManagerFor(event.singleProgram)%}
+                <td> <a href="/eventsList/{{event.id}}/edit" class=text-secondary>{{event.name}}</a></td>
+              {% else %}
+                <td> <a href="/eventsList/{{event.id}}/view" class=text-secondary>{{event.name}}</a></td>
+              {% endif %}
+              <td nowrap  class=text-secondary >{{event.startDate.strftime('%m/%d/%Y')}}</td>
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: Class values should be wrapped in quotations so it would be class="text-secondary".

File: app/templates/events/event_list.html [hunk 2]
```diff
@@ -52,9 +52,10 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
   </ul>
 </div>
 {% macro createTable(events, type) %}
+<p> checkthis{{ events|length }}</p>
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: What is the point of this?

File: app/templates/events/event_list.html [hunk 3]
```diff
@@ -69,22 +70,62 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
     </thead>
     <tbody>
         {% for event in events %}
+        {% set Fade_event= "" %}
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: We use camelCase everywhere for things like this so for consistency change this to fadeEvent.

File: app/logic/events.py [hunk 4]
```diff
@@ -446,3 +446,10 @@ def getTomorrowsEvents():
     tomorrowDate = date.today() + timedelta(days=1)
     events = list(Event.select().where(Event.startDate==tomorrowDate))
     return events
+def sortEvents(EventList):
```
Comment (AndersonStettner) [inline app/logic/events.py]: Lets try and make this name more descriptive. What events are being sorted or why are the events being sorted?

Comment (bledsoef) [inline app/logic/events.py]: I would reverse this sorting mechanism, it doesn't make much sense to interrupt the flow of the events list by changing the way are sorted. I would recommend having the newer past events towards the top and older ones further down like how active ones are sorted.

File: app/templates/events/event_list.html (hunk 2 above)
Comment (bledsoef) [inline app/templates/events/event_list.html]: I'm not sure if this was for testing but this should be removed before merge.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

**Review by bledsoef — CHANGES_REQUESTED** [review]
Looks good! Just a couple changes to make.

File: app/templates/events/event_list.html [hunk 5]
```diff
@@ -69,22 +69,22 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
     </thead>
     <tbody>
         {% for event in events %}
-          <tr>
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html:72]: Why did you remove this? Also, if it was intended to be deleted and it will not be put back make user you also remove its closing tag on line 106

File: app/templates/events/event_list.html [hunk 6]
```diff
@@ -104,9 +104,66 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
               </td>
             {% endif %}
           </tr>
+          {% endif %}
         {% endfor %}
      </tbody>
    </table>
+   {% if user.isCeltsStudentStaff or user.isAdmin %}
+   <button class="toggle-table text-primary border border-primary border-2">Past Events</button>
+ <table class="table d-none">
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: Make sure your indentation levels are matching and following what comes before it.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

**Review by AndersonStettner — CHANGES_REQUESTED** [review]
Also, I am not a big fan of how the "Past Events" button looks.
image

I think it would be best to have it look/styled like all the other buttons in the application.

File: app/templates/events/event_list.html (hunk 1 above)
Comment (bledsoef) [reply app/templates/events/event_list.html]: This still needs resolving @qasema

File: app/logic/events.py (hunk 4 above)
Comment (bledsoef) [reply app/logic/events.py]: This issue still needs resolved @qasema

**Review by AndersonStettner — CHANGES_REQUESTED** [review]
Instead of just hiding the "Past Events" button and showing an empty table in the accordion for students and faculty in the case where there are past events but no upcoming we should make sure to show the "There are no upcoming events for this term." message that appears when no events have occurred and there are none planned. 

What it looks like currently: 
image

What I think it should look like: 
image

**Comment by bledsoef** [comment]
If there are no upcoming events, show past events by default.

**Comment by bledsoef** [comment]
In our opinion we like the buttons looking like this:

Screenshot_20230301_092942


But you should ask for opinions via Slack.

File: app/templates/events/event_list.html [hunk 7]
```diff
@@ -69,22 +69,22 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
-              <td> <a href="/eventsList/{{event.id}}/edit" class="link-primary">{{event.name}}</a></td>
+              <td> <a href="/eventsList/{{event.id}}/edit" class=text-primary>{{event.name}}</a></td>
             {% else %}
-              <td> <a href="/eventsList/{{event.id}}/view" class="link-primary">{{event.name}}</a></td>
+              <td> <a href="/eventsList/{{event.id}}/view" class=text-primary>{{event.name}}</a></td>
             {% endif %}
-            <td nowrap>{{event.startDate.strftime('%m/%d/%Y')}}</td>
-            <td nowrap>{{event.timeStart.strftime('%-I:%M %p')}} - {{event.timeEnd.strftime('%-I:%M %p')}}</td>
-            <td>{{event.location}}</td>
+            <td nowrap  class=text-primary >{{event.startDate.strftime('%m/%d/%Y')}}</td>
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: You can't have text be blue unless it is clickable. In this case it is event more confusing since the event name is a link.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

**Comment by AndersonStettner** [comment]
Talk with either myself or Brian next time you come in about this issue.

**Comment by Karina-Agliullova** [comment]
I agree with Anderson and Finn. I think the Past Events button should be styled differently. And also, it would be better to have past events displayed by default when there are no upcoming events shown. It took me a while to understand that I need to click on the past events since there was nothing displayed. Other than that, good job!

**Comment by tylerpar99** [comment]
Looks good

**Review by AndersonStettner — CHANGES_REQUESTED** [review]
There are failing tests in `test_event_list.py`
Also, resolve merge conflicts.

**Comment by qasema** [comment]
[Issue 813 event list](https://trello.com/c/otizyOh5/607-issue-813-event-list)

File: app/templates/events/event_list.html [hunk 8]
```diff
@@ -71,10 +69,11 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
         <th scope="col"></th>
       </tr>
     </thead>
-    {% if not events[0].isPast %}
-    {% set pastEvents = "none" %}
     <tbody>
       {% for event in events %}
+      {% if events[events.index(event)] == events[0] and events[events.index(event)].isPast %}
+     <tr> No upcoming events for this term </tr>
```
Comment (Karina-Agliullova) [inline app/templates/events/event_list.html]: Fix indentation

File: app/templates/events/event_list.html [hunk 9]
```diff
@@ -51,36 +51,46 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
+  {% if not events[0].isPast  %}
+    {% set visibleEvent = "d-none" %}
+    {% set visibleClass = "showlist" %}
+  {% endif %}
+  <div class="table-responsive">
+    <table class="table ">
+      <thead>
+        <tr>
+          {% if type != "bonner" and type != "studentLed" %}
+          <th scope="col">Program</th>
```
Comment (thawpt) [inline app/templates/events/event_list.html]: fix indentation

File: app/templates/events/event_list.html [hunk 10]
```diff
@@ -51,36 +51,46 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
+          <th scope="col">Location</th>
+          <th scope="col"></th>
+        </tr>
+      </thead>
+      <tbody>
         {% for event in events %}
-          <tr>
-            {% if type != "bonner" %}
+        {% if events[events.index(event)] == events[0] and events[events.index(event)].isPast %}
+      <tr> No upcoming events for this term </tr>
```
Comment (Karina-Agliullova) [inline app/templates/events/event_list.html]: fix indentation

**Comment by tylerpar99** [comment]
Also the terminal prints hello when you visit some function

**Comment by tylerpar99** [comment]
Looks good!

**Comment by qasema** [comment]
[Issue 813 event list](https://trello.com/c/otizyOh5/607-issue-813-event-list)

File: app/templates/events/event_list.html [hunk 11]
```diff
@@ -53,137 +53,147 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
+                </td>
+              </tr>
             {% endif %}
-          </tr>
-        {% endfor %}
-     </tbody>
-   </table>
- </div>
+            {% if event.isPast %}
+              <tr class=" {{visibleClass}} table-secondary {{visibleEvent}}">
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: change `table-secondary` to `table-light`.

File: app/templates/events/event_list.html [hunk 12]
```diff
@@ -53,137 +53,147 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
-                <button type="button" class="btn btn-primary" onclick="rsvpForEvent({{event.id}})">RSVP</button>
+                </td>
               {% endif %}
-              </td>
+            </tr>
+          {% endif %}
+          {% if user.isCeltsStudentStaff or user.isAdmin %}
+            {% if not event.isPast and events[events.index(event)+1].isPast %}
+              <tr>
+                <td>
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: Remove the bottom border on this element. 

Also, instead of having a `Past Events` button change it to something a heading or title that you click to expand that part of the table. I thing the name should be change to something like `View Past Events` so it is clear what action is being done.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]
Past events should not be shown by default ever, even when there are no upcoming events. 

There is a messed up/missing closing or opening tag somewhere, there should be some margins around the table edges and the edge of the accordion like this:
table_margin

but your tables are missing that: 
image

File: app/templates/events/event_list.html (hunk 11 above)
Comment (qasema) [reply app/templates/events/event_list.html]: I thought that past events should be shown by default when there are no upcoming events. It was mentioned in a previous comment by Finn

**Comment by AndersonStettner** [comment]
[(Already in PR)On Events list page only show current and future events to students and gray out past events for Admin and Student Staff #813](https://trello.com/c/ArohxtxT/584-already-in-pron-events-list-page-only-show-current-and-future-events-to-students-and-gray-out-past-events-for-admin-and-student)

File: app/static/js/event_list.js [hunk 13]
```diff
@@ -41,3 +41,16 @@ function removeRsvpForEvent(eventID){
 
   })
 }
+
+document.getElementById("toggleButton").addEventListener("click", function() {
+  var tableRows = document.getElementsByClassName("showlist");
```
Comment (BrianRamsay) [inline app/static/js/event_list.js]: We should be doing this the jquery way like all of our other code, not the regular javascript way

File: app/static/js/event_list.js [hunk 14]
```diff
@@ -41,3 +41,16 @@ function removeRsvpForEvent(eventID){
...
+
+  for (var i = 0; i < tableRows.length; i++) {
+    var tableRow = tableRows[i];
+    if (tableRow.classList.contains("d-none")) {
+      tableRow.classList.remove("d-none");
+    } else {
+      tableRow.classList.add("d-none");
```
Comment (BrianRamsay) [inline app/static/js/event_list.js]: Use .hide() and .show() on tablerow instead of adding and removing `d-none` (you will need to use `display:none` in the html instead of the `d-none` class)

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/logic/events.py [hunk 15]
```diff
@@ -441,6 +441,13 @@ def getTomorrowsEvents():
     events = list(Event.select().where(Event.startDate==tomorrowDate))
     return events
 
+def upcomingEventsFirst(EventList):
+    """Sorts events so that upcoming events come first and past events come last"""
```
Comment (AndersonStettner) [inline app/logic/events.py]: This sort is not happening on the Student Led Service tab.

File: app/templates/events/event_list.html [hunk 16]
```diff
@@ -28,6 +28,15 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
       </div>
     </div>
   </div>
+  {% if user.isCeltsStudentStaff or user.isAdmin %}
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: `.isAdmin` is a property in the User model that is both `isCeltsAdmin` and `isCeltsStudentStaff` so the check for `isCeltsStudentStaff` here should not be needed.

File: app/templates/events/event_list.html [hunk 17]
```diff
@@ -53,139 +62,134 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
-  <table class="table table-striped">
-    <thead>
-      <tr>
-        {% if type != "bonner" %}
-        <th scope="col">Program</th>
+  <div class="accordion-body table-responsive">
+    <table class="table ">
+      <thead>
+        <tr>
+          {% if type != "bonner" and type != "studentLed" %}
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: Why have you added `"studentLed"` to this condition?

**Review by AndersonStettner — CHANGES_REQUESTED** [review]
When there are no upcoming events in the Student Led Service tab, there is no message about the tab, it is just empty space: 
image

I have an upcoming event but I still have the "No upcoming events" status message:
image

If there are past events but no future events we should not show the empty table under the "No future events" message.

File: app/templates/events/event_list.html (hunk 17 above)
Comment (Mercy-Eze) [reply app/templates/events/event_list.html]: We didn't add studentLed to the condition. We left it as was in development and it seemed to work fine

File: app/logic/events.py (hunk 15 above)
Comment (BrianRamsay) [reply app/logic/events.py]: I think we talked about not using `upcomingEventsFirst` at all, on any of them. just sort in chronological order is fine (which is done in the query already)

File: app/static/js/event_list.js [hunk 18]
```diff
@@ -5,8 +5,30 @@ $(document).ready(function(){
     $("#rsvpBtn").click(function(){
         rsvpForEvent($("#rsvpBtn").val())
     })
+    $("#toggleButton").on("click", toggleButton)
```
Comment (BrianRamsay) [inline app/static/js/event_list.js]: What does this do? There is no `toggleButton` function to be used as an event handler

File: app/static/js/event_list.js [hunk 19]
```diff
@@ -5,8 +5,30 @@ $(document).ready(function(){
...
+    localStorage.removeItem("toggleState");
```
Comment (BrianRamsay) [inline app/static/js/event_list.js]: If you remove this when the page loads, how do you use it to decide whether to check the checkbox or not?

File: app/templates/events/event_list.html [hunk 20]
```diff
@@ -28,6 +28,15 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
       </div>
     </div>
   </div>
+  {% if user.isCeltsStudentStaff or user.isAdmin %}
+    <div class="row px-4">
+      <div class = "d-none d-sm-block col "></div>
+      <div class="form-check form-switch mb-2 col-auto">
+        <label class="form-check-label" for="toggleButton">View Past Events</label>
+        <input class="form-check-input" type="checkbox" id="toggleButton" autocomplete="off">
```
Comment (BrianRamsay) [inline app/templates/events/event_list.html]: `toggleButton` needs a more specific name, like `viewPastEventsToggle` or `viewPastEventsButton`

File: app/templates/events/event_list.html [hunk 21]
```diff
@@ -53,139 +62,134 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
+                  </td>
+                {% else %}
+                  <td>
+                  {% if event.id in rsvpedEventsID %}
+                    <button type="button" class="btn btn-danger" onclick="removeRsvpForEvent({{event.id}})">Remove RSVP</button>
+                  {% elif (not event.isPast) and (event.isRsvpRequired) %}
+                    <button type="button" class="btn btn-primary" onclick="rsvpForEvent({{event.id}})">RSVP</button>
+                  {% endif %}
+                  </td>
+                {% endif %}
```
Comment (BrianRamsay) [inline app/templates/events/event_list.html]: We shouldn't duplicate all of this code for showing the rows

File: app/templates/events/event_list.html [hunk 22]
```diff
@@ -53,139 +62,134 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
+            {% if event.isPast %}
+              <tr class="table-light showlist" style="display: none;">
+                {% if type != "bonner" and type != "studentLed" %}
+                  {% if event.singleProgram%}
+                    <td  class="text-secondary">{{event.singleProgram.programName}}</td>
+                  {% else %}
+                    <td  class="text-secondary">This event does not belong to any program.</td>
+                  {% endif %}
+                {% endif %}
+                {% if user.isCeltsAdmin or user.isProgramManagerFor(event.singleProgram)%}
```
Comment (BrianRamsay) [inline app/templates/events/event_list.html]: Did this come from a merge conflict? `event.singleProgram` doesn't exist anymore, and should break in your code.

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

File: app/static/js/event_list.js (hunk 18 above)
Comment (Mercy-Eze) [reply app/static/js/event_list.js]: Oversight: ...was used when a toggle button function was defined as a separate function

File: app/templates/events/event_list.html [hunk 23]
```diff
@@ -68,31 +77,49 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
...
+              {% if event.isPast %}
+                <td> <a href="/event/{{event.id}}/view" class="link-secondary">{{event.name}}</a></td>
+              {% else %}
+                <td> <a href="/event/{{event.id}}/view" class="link-primary">{{event.name}}</a></td>
+              {% endif %}
             {% endif %}
             <td nowrap>{{event.startDate.strftime('%m/%d/%Y')}}</td>
             <td nowrap>{{event.timeStart.strftime('%-I:%M %p')}} - {{event.timeEnd.strftime('%-I:%M %p')}}</td>
-            <td>{{event.location}}</td>
+            <td class="text-dark">{{event.location}}</td>
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html:102]: You should `set` `primary` or `secondary` depending on the case you want, then we wont need the copied lines.

File: app/templates/events/event_list.html (hunk 21 above)
Comment (AndersonStettner) [reply app/templates/events/event_list.html]: Like the above section, use `set` to change to the value that you want.

File: app/templates/events/event_list.html [hunk 24]
```diff
@@ -105,14 +132,20 @@ <h1 class="text-center">Events List for {{selectedTerm.description}}</h1>
               </td>
             {% endif %}
           </tr>
-        {% endfor %}
+      {% endfor %}
+      {% if selectedTerm == g.current_term and events[-1].isPast %}          
+        <tr class>
```
Comment (AndersonStettner) [inline app/templates/events/event_list.html]: Remove this class if you do not need to use it.

**Review by AndersonStettner — CHANGES_REQUESTED** [review]

**Review by AndersonStettner — CHANGES_REQUESTED** [review]
If there is no upcoming event and we are not displaying past events we should not show the empty table, we should just show the "There are no events" message. 

Also, in the case where there are no upcoming events and we are showing past events the bottom line under "There are no events" message does not extend to the end of the table, this should be fixed. 
image

I thought that if we looked at events from a past term the `View Past Events` toggle would be set by default? If that is not the case currently it should be. This is what it looks like right now if you go to look at events in a past term: 
image

If a program is marked as "Student Led" we should not display the program in the table. The table is already inside an accordion with the name of the program as its title. 
image

File: app/static/js/event_list.js [hunk 25]
```diff
@@ -5,6 +5,34 @@ $(document).ready(function(){
...
+  
+
+    viewPastEventsToggle.prop("checked", false);
+    toggleRows(false);  
+
+    viewPastEventsToggle.prop("checked", g_isPastTerm);
+    if (g_isPastTerm){
+      toggleRows(true)
+      viewPastEventsToggle.prop("disabled", true);
+    }
```
Comment (BrianRamsay) [inline app/static/js/event_list.js]: Lines 11-18 can be replaced with just the prop() lines and one toggleRows call all using `g_isPastTerm`, because you want to pass or set true for all of them when `g_isPastTerm` is true, and false when it is false.

File: app/static/js/event_list.js [hunk 26]
```diff
@@ -5,6 +5,34 @@ $(document).ready(function(){
...
+      toggleRows(isChecked);
+
+      localStorage.setItem("toggleState", isChecked ? "checked" : "unchecked")
+    });
+
+    function toggleRows(isChecked) {
+      var tableRows = $(".showlist");
+      if (isChecked){
+        tableRows.show();
+      }
```
Comment (BrianRamsay) [inline app/static/js/event_list.js]: let's keep the `else` on the same line as the end brace, according to our coding style

**Review by BrianRamsay — CHANGES_REQUESTED** [review]

**Review by BrianRamsay — APPROVED** [review]
